### PR TITLE
Adds ability to disable indexing for v2 storage implementations

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This starts towards "firehose mode" by making it possible to disable
indexing, notably on cassandra where SASI amplifies writes into the
cluster.

When enabled, only retrieval by ID is allowed. That said, someone can
feel free to run the dependency linker if they like.. If 100% sampling
is on, the dependency links will relate to actual service calls made.

See #1869

### CODE REVIEW

**Review by michaelsembwever — APPROVED** [review]

**Comment by michaelsembwever** [comment]
I notice there's no check that the SASIs have _actually_ been dropped.
With `indexing=false` do you still want the writes to be getting amplified (which will happen unless the SASI are dropped via `cql`)?

**Comment by codefromthecrypt** [comment]
@michaelsembwever yeah incomplete on disabling things for CQL. I'm not sure we will check at runtime, just install without SASI on a fresh spot. We could check though

**Comment by codefromthecrypt** [comment]
renamed to "search-enabled" as seems to imply more what's being disabled. feedback welcome

**Comment by codefromthecrypt** [comment]
@igorwwwwwwwwwwwwwwwwwwww if /config.json has searchEnabled=false, you think you could hide the find screen?

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
@adriancole that would be hiding the annotations query input, right? should be quite straight forward to change in zipkin-ui.

**Comment by drolando** [comment]
@adriancole This is awesome! It'd have probably taken me a week to find all the places I needed to change :)

**Comment by codefromthecrypt** [comment]
added a commit to not install the indexes unless search is enabled. Since a lot of prod sites will have their own copies of the schema files, they'll have to do the same on their own I guess. That or we further templatize the schema files with things like replication factor and ttl cc @openzipkin/cassandra

**Comment by drolando** [comment]
I don't think you can add the replication factor in your schema since you've no idea about the datacenters names.

**Comment by codefromthecrypt** [comment]
@drolando thx for the hint.. will not be ambitious then

**Comment by codefromthecrypt** [comment]
added elasticsearch implementation. there are a few areas we need some unit tests, and generally some docs. We also need to have the UI conditional on the searchEnabled property

File: zipkin-storage/zipkin2_cassandra/src/main/java/zipkin2/storage/cassandra/CassandraSpanStore.java [hunk 1]
```diff
@@ -85,7 +99,8 @@
     final int traceIndexFetchSize = request.limit() * indexFetchMultiplier;
     List<Call<Map<String, Long>>> callsToIntersect = new ArrayList<>();
 
-    List<String> annotationKeys = CassandraUtil.annotationKeys(request);
```
Comment (codefromthecrypt) [inline zipkin-storage/zipkin2_cassandra/src/main/java/zipkin2/storage/cassandra/CassandraSpanStore.java]: self-note: scrub this as the call is obviated higher up

File: zipkin-server/README.md [hunk 2]
```diff
@@ -102,6 +102,7 @@ Attribute | Property | Description
 --- | --- | ---
 environment | zipkin.ui.environment | The value here becomes a label in the top-right corner. Not required.
 defaultLookback | zipkin.ui.default-lookback | Default duration in millis to look back when finding traces. Affects the "Start time" element in the UI. Defaults to 3600000 (1 hour in millis).
+searchEnabled | zipkin.ui.search-enabled | If the Find Traces screen is enabled. Defaults to true.
```
Comment (drolando) [inline zipkin-server/README.md:105]: we also need to say something about why you'd ever want to do this

Comment (codefromthecrypt) [reply zipkin-server/README.md:105]: PTAL cc also @bplotnick

**Comment by codefromthecrypt** [comment]
current status: revised server README and rebased

still a few areas we need some unit tests, and generally some docs. We also need to have the UI conditional on the searchEnabled property

**Comment by codefromthecrypt** [comment]
@igorwwwwwwwwwwwwwwwwwwww @Logic-32 can I ask one of you to figure out how to hide the "find a trace" and also maybe change the default content to empty or otherwise when the config.json has searchEnabled=false? I could possibly figure this out, but it will likely take orders of magnitude more time. I'll offer a trade to implement a favorite issue, like https://github.com/openzipkin/brave/issues/564 for example :D

**Comment by codefromthecrypt** [comment]
pretend you don't know I'll implement the above issue anyway

**Comment by drolando** [comment]
@adriancole do you want to update the UI as part of this CR? It feels long enough already :)

**Comment by codefromthecrypt** [comment]
>
> @adriancole <https://github.com/adriancole> do you want to update the UI
> as part of this CR? It feels long enough already :)
>
well I like to keep master shippable. So, this question means.. is hiding
the UI an optional part of this change? Since you are the customer of this
change, if you say so, works for me!

**Comment by codefromthecrypt** [comment]
So, plan is to "soft launch" this so that @drolando and others can try it and give feedback eventhough the UI won't be disabled. I'll do that as a patch release. We'll make a more broadcast announcement once feedback comes in (and any changes along with it), and ideally UI support is also in.

**Comment by drolando** [comment]
> well I like to keep master shippable

Yeah, I agree with having a working master. But I usually tend to do multiple smaller PRs that change it gradually to not have a too long testing and reviewing cycle (and huge diffs). :)

You're the main maintainer though, so I'm fine with whatever you prefer.

**Comment by codefromthecrypt** [comment]
ok tests are backfilled. I only need to write relevant README sections in cassandra and elasticsearch

**Comment by codefromthecrypt** [comment]
ok all done with tests and README will merge on green

**Comment by codefromthecrypt** [comment]
opened a follow-up issue for the UI change https://github.com/openzipkin/zipkin/issues/1893

**Pull request merged**